### PR TITLE
various improvements to HLT-development tools

### PR DESCRIPTION
**Pull request by missirol**
#### PR description:

This PR contains various improvements to scripts used for HLT development (used both by TSG itself and HLT developers at large).

The main changes are
 1. removal of `ConfDB` queries from the so-called HLT-Validation tests running in IBs (to address #37598 and #39345);
 2. make more HLT-development scripts capable of using the `--dbproxy` option, to connect to `ConfDB` from outside CERN through a tunnel connection.

As part of 1., scripts like `hltIntegrationTests` and `hltListPaths` are extended to be able to use as input a local cfg file (as an alternative to the standard HLT-menu input, which is the name of a configuration in the `ConfDB` database).

Other smaller updates are included in the process. More detailed comments will be added as part of the review.

Summary:
```bash
./getHLT.sh --dbproxy # will now work
./getFrozenHLT.sh --dbproxy # will now work
./makeSubtables --dbproxy # will now work
./runAll.csh [..] # does not query ConfDB anymore (content of the tests is unchanged)

hltIntegrationTests hlt.py [..] # will now work
hltListPaths hlt.py [..] # will now work
```

The aim is to add functionalities, and leave current behaviour unchanged (except for tweaks to the printouts of `hltIntegrationTests` and `get{Frozen,}HLT.sh`).

Merely technical. No differences expected in the outputs of PR tests (this PR is basically insensitive to PR tests).

#### PR validation:

Private tests.

#### If this PR is a backport, please specify the original PR and why you need to backport that PR. If this PR will be backported, please specify to which release cycle the backport is meant for:

`CMSSW_12_4_X`
`CMSSW_12_5_X`

(Backports are not necessary, but they could be useful to keep these tools aligned across release cycles.)

### CODE REVIEW

File: HLTrigger/Configuration/scripts/hltCheckPrescaleModules [hunk 1]
```diff
@@ -17,7 +15,12 @@ if len(sys.argv) != 2:
   sys.exit(1)
 
 # whitelist paths exempt from validation
-whitelist = ('HLTriggerFirstPath', 'HLTriggerFinalPath')
+whitelist = [
+  'HLTriggerFirstPath',
+  'HLTriggerFinalPath',
+  'Status_OnCPU',
+  'Status_OnGPU',
```
Comment (missirol) [inline HLTrigger/Configuration/scripts/hltCheckPrescaleModules:22]: This removes these expected errors:
```
ERROR: path Status_OnCPU dos not have an HLTPrescaler module
ERROR: path Status_OnGPU dos not have an HLTPrescaler module
```
https://cmssdt.cern.ch/SDT/jenkins-artifacts/HLT-Validation/CMSSW_12_6_X_2022-11-05-1100/el8_amd64_gcc10/HLT_Integration_GRun_MC.log

File: HLTrigger/Configuration/scripts/hltGetConfiguration [hunk 2]
```diff
@@ -243,8 +242,19 @@ parser.add_argument('--help',
...
+  if skipNext:
+    skipNext = False
+    continue
+  if cmdArg.startswith('--dbproxy'):
+    if cmdArg.startswith('--dbproxyh') or cmdArg.startswith('--dbproxyp'):
+      skipNext = '=' not in cmdArg
+    continue
+  cmdArgs += [cmdArg]
+cmdArgs[0] = os.path.basename(cmdArgs[0])
+
```
Comment (missirol) [inline HLTrigger/Configuration/scripts/hltGetConfiguration:257]: This avoids having `--dbproxy` in the 1st line of the output of `hltGetConfiguration`. The db-proxy option should arguably be "untracked", so to speak. Plus, without this we would have spurious diffs in the first line of TSG menus dumped in the release (depending on how `getHLT.sh` was run).

File: HLTrigger/Configuration/scripts/hltIntegrationTests [hunk 3]
```diff
@@ -273,46 +348,63 @@ process.HLTriggerFinalPath = cms.Path( process.hltTriggerSummaryAOD + process.hl
...
+# set allowed accelerators
+process.options.accelerators = [ "$ACCELERATOR" ]
 
 process.hltTriggerSummaryAOD.throw = cms.bool( True )
 @EOF
 
-# dump the menu name, and its release template
-log "HLT menu: $(head -n1 hlt.py | cut -c 3-)"
+# find the list of all trigger paths
+TRIGGERS=$(hltListPaths hlt.py -p --no-dep --exclude "^HLTriggerFinalPath$")
```
Comment (missirol) [inline HLTrigger/Configuration/scripts/hltIntegrationTests:380]: Using the python file as input to `hltListPaths` allows us to support the following (unlike in the past)
```bash
# run integration tests only on Ele* Paths of GRun
hltIntegrationTests /dev/CMSSW_12_4_0/GRun -x --paths "HLT_Ele*_v*" [..]
```

File: HLTrigger/Configuration/scripts/hltIntegrationTests [hunk 4]
```diff
@@ -28,66 +27,76 @@ Usage:
...
+  -d | --dir         WORKDIR      Create all files and run all tests inside WORKDIR (defauls: ./hltintegration)
+  -i | --input       INPUT        Use the specified EDM file as input (data tier: RAW)
+  -n | --size        EVENTS       Run on EVENTS events (-1 for all, default is 100)
+  -k | --skip        EVENTS       Skip the first EVENTS (default is 0)
+  -e | --events      EVENTS       Run on a comma-separated list of EVENTS, a VEventRange
+  -j | --jobs        JOBS         Run JOBS single trigger jobs in parallel (default 4)
+       --streams     STREAMS      Run with STREAMS parallel streams (i.e. events) (default 0 means as many streams as threads)
+       --threads     THREADS      Run with THREADS threads when running the whole HLT (default 4)
+  -a | --accelerator ACCELERATOR  Keyword to choose allowed accelerators (examples: \"*\", \"cpu\", \"gpu-nvidia\")
+  -x | --extra       OPTIONS      If the HLT menu is a local cmsRun cfg file, OPTIONS is used as
```
Comment (missirol) [inline HLTrigger/Configuration/scripts/hltIntegrationTests:45]: If the input of the integration tests is a cfg file, the `-x` option can be used to specify command-line arguments for when cmsRun runs that cfg, e.g. `hltIntegrationTests hlt.py -x realData=0 [..]` will lead to `cmsRun hlt.py realData=0 [..]`.

This feature is used in `runIntegration.csh`, to use the `OnLine` dumps for the HLT-Validation tests without having to craft new config files.

File: HLTrigger/Configuration/scripts/hltListPaths [hunk 5]
```diff
@@ -31,74 +36,88 @@ def getPathList(config):
...
-    pathDict = process.finalpaths_()
+    useFinalPaths = True
 
   # Paths, EndPaths, and FinalPaths ('all')
+  elif config.selection == 'all':
+    usePaths, useEndPaths, useFinalPaths = True, True, True
+
+  # invalid value
   else:
-    pathDict = zip(process.paths_(), process.endpaths_(), process.finalpaths_())
```
Comment (missirol) [inline HLTrigger/Configuration/scripts/hltListPaths:48]: `zip` here was not doing what I was naively expecting (`hltListPaths -a` can return incorrect results without this PR).

File: HLTrigger/Configuration/test/cmsDriver.csh [hunk 6]
```diff
@@ -1,8 +1,6 @@
-#! /bin/tcsh
+#!/bin/tcsh
 
-cmsenv
-
-rehash
+eval `scram runtime -csh`
```
[the remark this answers is not in the document]
Comment (missirol) [reply HLTrigger/Configuration/test/cmsDriver.csh:3]: I would like to ask a question to @smuzaffar .

In https://github.com/cms-sw/SCRAM/blob/a5e2033a2ed7f6b84d6fd92a386a920f02bdb54b/ReleaseNotes.txt#L226 it sounds like running `rehash` is not necessary after doing `eval 'scram runtime -csh'`. Is that correct, Shahzad ?

File: HLTrigger/Configuration/test/getEventContent.py [hunk 7]
```diff
@@ -1,35 +1,58 @@
-#! /usr/bin/env python3
```
Comment (missirol) [inline HLTrigger/Configuration/test/getEventContent.py:1]: `getEventContent.py`:
 - added `argparse` and support for `--dbproxy` option
 - minimised number of ConfDB queries (from 4 to 1)
 - output unchanged, but now the scripts outputs to stdout (not to a file whose name is hard-coded inside the script)

File: HLTrigger/Configuration/test/runIntegration.csh [hunk 8]
```diff
@@ -28,44 +27,34 @@ endif
...
-    else if ( ($table == Fake1) || ($table == Fake2) || ($table == 2018) ) then
-      set basegt = auto:run2_${infix}_${table}
-    else
-      set basegt = auto:run3_${infix}_${table}
-    endif
-    set autogt = "--globaltag=${basegt}"
     set infile = file:../RelVal_Raw_${table}_${gtag}.root
 
-#   -x "--l1-emulator" -x "--l1 L1GtTriggerMenu_L1Menu_Collisions2012_v1_mc" 
+    set hltIntegTestCmd = "hltIntegrationTests OnLine_HLT_${table}.py ${extraflags} -d ${name} -i ${infile} -n 100 -j 4 -a cpu"
```
Comment (missirol) [inline HLTrigger/Configuration/test/runIntegration.csh:44]: This changes the input of the hltIntegrationTests that run in IBs, in such a way that ConfDB is not queried anymore.

File: HLTrigger/Configuration/test/cmsDriver.csh (hunk 6 above)
[the remark this answers is not in the document]
Comment (smuzaffar) [reply HLTrigger/Configuration/test/cmsDriver.csh:3]: @missirol , yes this is correct, there is no need to run `rehash`

[the remark this answers is not in the document]
Comment (missirol) [reply HLTrigger/Configuration/test/cmsDriver.csh:3]: Okay, thanks @smuzaffar .

File: HLTrigger/Configuration/test/runIntegration.csh [hunk 9]
```diff
@@ -28,44 +27,34 @@ endif
...
 
-    set config = `grep tableName OnLine_HLT_${table}.py | cut -f2 -d "'"`
-    if ($table == Fake) then
-      set basegt = auto:run1_${infix}_${table}
-    else if ( ($table == Fake1) || ($table == Fake2) || ($table == 2018) ) then
-      set basegt = auto:run2_${infix}_${table}
-    else
-      set basegt = auto:run3_${infix}_${table}
-    endif
-    set autogt = "--globaltag=${basegt}"
```
Comment (Martin-Grunewald) [inline HLTrigger/Configuration/test/runIntegration.csh:53]: Where is the globaltag now constructed and set for the integration tests?

**Comment by Martin-Grunewald** [comment]
Given ```getHLT.sh``` is changed, I would expect a change also in ```getFrozenHLT.sh```.

File: HLTrigger/Configuration/test/runIntegration.csh (hunk 9 above)
Comment (missirol) [reply HLTrigger/Configuration/test/runIntegration.csh:53]: `globalTag=@` resolves to the GT hard-coded in [`OnLine_HLT_GRun`](https://raw.githubusercontent.com/cms-sw/cmssw/master/HLTrigger/Configuration/test/OnLine_HLT_GRun.py), the data or the MC one depending on the value of `realData`. This follows what is currently done in the [`addOnTests`](https://github.com/cms-sw/cmssw/blob/d38f357970f97327065d03b5d5f5a5a9b9cc10e7/Utilities/ReleaseScripts/scripts/addOnTests.py#L98).

Comment (Martin-Grunewald) [reply HLTrigger/Configuration/test/runIntegration.csh:53]: Ah great, removes this duplication then!

**Comment by missirol** [comment]
> Given `getHLT.sh` is changed, I would expect a change also in `getFrozenHLT.sh`.

Indeed, I need to add this.

**Comment by missirol** [comment]
The latest version includes the update of `getFrozenHLT.sh` (thanks for spotting this, Martin).

I still have a couple of small improvements in mind, and more testing to do. The plan would be to integrate this PR and its (future) backports next week.

**Comment by missirol** [comment]
please test

The latest push updated `get{Frozen,}HLT.sh` such that they can be executed from any directory, not just `test/`. I don't have other updates planned for this PR (and more testing didn't reveal issues). I'd backport this down to `12_4_X`, mainly because those are the cycles with frequent IBs, so most affected by #37598.

**Pull request merged**